experimental/iterators/test/Dialect/Iterators/iterator-states-invalid.mlir

```diff
@@ -0,0 +1,75 @@
+// Test that we can parse and verify iterator states without errors
+// RUN: mlir-proto-opt --verify-diagnostics --split-input-file %s
+
+!sampleInputState = type !iterators.sampleinputstate<i32>
+func @testCreateSampleInputTypeMismatch() {
+  // expected-error@+1 {{Sample input iterator (currently) has to return elements of type 'i32'}}
+  %initialState = "iterators.createSampleInputState"() : () -> !iterators.sampleinputstate<i64>
+  return
+}
+
+// -----
+
+!sampleInputState = type !iterators.sampleinputstate<i32>
+!wrongSampleInputState = type !iterators.sampleinputstate<i64>
+!wrongReduceState = type !iterators.reducestate<!wrongSampleInputState>
+func @testCreateSampleInputTypeMismatch() {
+  %initialUpstreamState = "iterators.createSampleInputState"() : () -> !sampleInputState
+  // expected-error@+1 {{Upstream iterator of reduce iterator must produce elements of type 'i64' but produces elements of type 'i32'}}
+  %initialState = "iterators.createReduceState"(%initialUpstreamState) : (!sampleInputState) -> !wrongReduceState
+  return
+}
+
+// -----
+
+!sampleInputState = type !iterators.sampleinputstate<i32>
+func @testOpenElementTypeMismatch() {
+  %initialState = "iterators.createSampleInputState"() : () -> !sampleInputState
+  // expected-error@+1 {{Opening iterator of type '!iterators.sampleinputstate<i32>' should return the same type but returns '!iterators.sampleinputstate<i64>'}}
+  %openedState = "iterators.open"(%initialState) : (!sampleInputState) -> !iterators.sampleinputstate<i64>
+  return
+}
+
+// -----
+
+!sampleInputState = type !iterators.sampleinputstate<i32>
+!reduceState = type !iterators.reducestate<!sampleInputState>
+func @testOpenStateTypeMismatch() {
+  %initialState = "iterators.createSampleInputState"() : () -> !sampleInputState
+  // expected-error@+1 {{Opening iterator of type '!iterators.sampleinputstate<i32>' should return the same type but returns '!iterators.reducestate<!iterators.sampleinputstate<i32>>'}}
```

Comment:
This message and other similar ones in the file are generally considered "standard Tablgen do not reinvent".
Look around `SameOperandsElementType` in OpBase.td and see if you find something that suits you.

Reply:
Aha, this is a super interesting pointer! `SameOperandsElementType` only works on [`ShapedType`](https://mlir.llvm.org/docs/Dialects/Builtin/#shapedtype-shapedtypeinterface) (due to how `getElementTypeOrSelf` is [implemented](https://github.com/llvm/llvm-project/blob/09c2b7c35af8c4bad39f03e9f60df8bd07323028/mlir/lib/IR/TypeUtilities.cpp#L24)), but actually maybe `Iterators` *are*/*should be* `ShapedType`s. If that were the case, we might get some inference facilities etc. for free, which may be a great example of why implementing database-style computations in MLIR is a good idea. Definitely, Iterators have an element type; they also have a rank, which is always `1` in all cases I am currently considering; and they have a shape, which is mostly unknown statically, but not always. I am not sure, though, if a `ShapedType` can decide only to work for rank `1`. What should my `clone` method do if it is invoked with a shape of rank `!= 1`?

Reply:
The more I think about this, the more I think that `ShapedType` may model too broad a concept: it models things that have elements *and* things that have a shape. Currently, I only need the former. There are `ContainerType` and `ShapedContainerType`, which do make that distinction, but all useful predicates (`SameOperandsElementType`) are based on `ShapedType` instead of those.

Reply:
Maybe I understand why: `ContainerType` is just a `class`, not an interface, so unlike `ShapedType` you can `dyn_cast` to it...

Reply:
Maybe what I propose can't be done: I am thinking of `ShapedType` (an interface which forces elements and a shape) that *is a* `Container` (which only forces elements, where `ShapedType` inherits all element-related properties from `Container`. I am not sure, but I think that that kind of inheritance it not supported with `OpInterface`...

Bottom line: it seems like I need to replicate `SameOperandsElementType` and similar for my purposes. Not very satisfying, TBH...

Reply:
Offline discussion: yes, I need to roll my own thing.